microhapulator/filter.py

```diff
@@ -64,37 +89,31 @@ def keep(self, r1, r2):
     def is_ambiguous(self, read):
         return read.seq.count("N") / len(read.seq) > self.threshold
 
-    def write_counts_output(self):
-        with self.outfiles["counts"] as counts_out:
-            header = f"R1Only\tR2Only\tR1andR2\tPairsRemoved\tPairsKept"
-            total_pairs_filtered = self.num_r1_failed + self.num_r2_failed + self.num_both_failed
-            counts_str = f"{self.num_r1_failed}\t{self.num_r2_failed}\t{self.num_both_failed}\t{total_pairs_filtered}\t{self.num_pairs_passed}"
-            counts_out.write(f"{header}\n{counts_str}\n")
-
-    def get_output_files(self, out_prefix):
-        self.outfiles = dict()
-        self.outfiles["r1"] = mhopen(f"{out_prefix}-ambig-filtered-R1.fastq", "w")
-        self.outfiles["r2"] = mhopen(f"{out_prefix}-ambig-filtered-R2.fastq", "w")
-        self.outfiles["r1_mates"] = mhopen(f"{out_prefix}-ambig-R1-mates.fastq", "w")
-        self.outfiles["r2_mates"] = mhopen(f"{out_prefix}-ambig-R2-mates.fastq", "w")
-        self.outfiles["counts"] = mhopen(f"{out_prefix}-ambig-read-counts.txt", "w")
+    @property
+    def summary(self):
+        header = "R1OnlyFailed\tR2OnlyFailed\tPairFailed\tPairsRemoved\tPairsKept\tTotalPairs"
+        count_data = f"{self.counts.r1_failed}\t{self.counts.r2_failed}\t{self.counts.pairs_failed}\t{self.counts.total_pairs_filtered}\t{self.counts.pairs_passed}\t{self.counts.total_pairs}"
+        return f"{header}\n{count_data}"
 
 
 class SingleReadFilter:
-    def __init__(self, reads_in, output_prefix):
-        self.reads_in = mhopen(reads_in, "r")
-        self.get_output_files(output_prefix)
+    def __init__(self, reads_in, reads_out):
+        self.reads_in = mhopen(reads_in, "r", enter=False)
+        self.reads_out = mhopen(reads_out, "w", enter=False)
```

Comment:
True, we don't need it for the output stream.

My changes to the `mhopen` function were to avoid opening four file handles with a single `with` statement. There's just no clean way to do that. I thought the changes had given us the best of both worlds—my testing of the workflow was certainly successful, but I clearly did some violence to the test suite.

I'm open to having some of these changes reverted or abandoned. But somehow we need a way to manage lots of file handles cleanly and responsibly (i.e. close them).

Reply:
Was your testing of the workflow for single ended reads successful too? In my testing, paired end reads were successful but single end was not. The implementation of the `PairedReadOutputStream` class allowed us to get rid of the `with` statement for 4 file handles.  These files aren't using `mhopen` though.  It seems that files opened with `mhopen` still need to use the `with` syntax to be managed.

Reply:
Hmm, yeah, that could be it. I was not as thorough with single end data testing.

Reply:
I think because of the `@contextmanager` decorator with `mhopen`, it is always returning a context manager.  I propose reverting the changes of `mhopen` and using it to handle the input files using `with` syntax.  And retaining the changes that use the `__init__` and `__del__` methods to handle the opening/closing of the output files. With this, we won't be opening more than 2 files opened using `with` syntax and will still be properly closing all input and output files correctly.

Reply:
I support that approach in principle. You're welcome to push to this branch, open a new branch, or just work on the original branch—whichever is most convenient.

Reply:
I made the changes on this branch, as well as updated the tests and merged it into the `filter-ambiguous` branch.